### PR TITLE
Add a keyboard shortcut to focus the global search bar (CMD + K/CTRL + K)

### PR DESCRIPTION
**Pull request by janfaracik**
Small PR that adds a keyboard shortcut for the search function in Jenkins. Press CMD + K (or CTRL + K) to quickly focus the global search bar. 

image

The advantage this brings is that it makes it faster to use the search function/encourages the use of it.

---

_Why CMD + K?_

It's quite a common shortcut used across various services, such as GitHub and Spotify. I'm planning on using `/` for in page search fields, such as the filter build history input or the plugin manager search input.

---

This is a small subset of the changes demoed in [May's UX meeting](https://youtu.be/amvQE0_nVBM?list=PLN7ajX_VdyaOnsIIsZHsv_fM9QhOcajWe&t=305). The eventual idea would be to extend the API to allow plugin developers to use keyboard shortcuts.

I've split it out for two reasons: 

1. Easier to review
2. There's some work around search/[a command palette](https://youtu.be/SxSObgwjwt8?list=PLN7ajX_VdyaOnsIIsZHsv_fM9QhOcajWe&t=3771) I'm putting together that would benefit from having this in3. 

---

✅ ~Would be good to have this tested on non-macOS systems~ Cheers @NotMyFault 

### Proposed changelog entries

* Keyboard shortcut added to focus global search bar (CMD + K/CTRL + K).

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [ ] (If applicable) Jira issue is well described
- [ ] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change) and are in the imperative mood. [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [ ] Appropriate autotests or explanation to why this change has no tests
- [ ] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadoc, as appropriate. 
- [ ] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")` if applicable.
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@jenkinsci/sig-ux 

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are accurate, human-readable, and in the imperative mood
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).

<a href="https://gitpod.io/#https://github.com/jenkinsci/jenkins/pull/6893"></a>

### CODE REVIEW

File: war/src/main/js/keyboard-shortcuts.js
Comment (NotMyFault) [inline war/src/main/js/keyboard-shortcuts.js:21]: Isn't platform detection that way deprecated?

Comment (janfaracik) [reply war/src/main/js/keyboard-shortcuts.js]: I seem to recall reading that (possibly on MDN) but it doesn't appear to mention that anymore, in fact the only time it recommends using `navigation.platform` is for our exact situation - https://developer.mozilla.org/en-US/docs/Web/API/Navigator/platform#examples

Comment (NotMyFault) [reply war/src/main/js/keyboard-shortcuts.js]: If MDN is fine with it we can and should use it for sure.

**Review by NotMyFault — APPROVED** [review]
Looks nice, tested on my surface to warrant Windows functionality too.

File: war/src/main/js/keyboard-shortcuts.js
Comment (daniel-beck) [inline war/src/main/js/keyboard-shortcuts.js:17]: This doesn't actually describe what this does correctly.

Also, is this intended to be reused elsewhere? Otherwise this is unnecessarily generic.

Comment (janfaracik) [reply war/src/main/js/keyboard-shortcuts.js]:
```js
/**
 * Given a keyboard shortcut, e.g. CMD+K, replace any included modifier keys for the user's
 * platform e.g. output will be CMD+K for macOS/iOS, CTRL+K for Windows/Linux
 * @param {string} keyboardShortcut The shortcut to translate
 */
function translateModifierKeysForUsersPlatform(keyboardShortcut) {
  const useCmdKey = navigator.platform.toUpperCase().indexOf("MAC") >= 0 ||
    navigator.platform.toUpperCase() === "IPHONE"
  return keyboardShortcut.replaceAll(/CMD|CTRL/ig, useCmdKey ? "CMD" : "CTRL")
}
```

I've updated it now to be more descriptive, hope that's better. 

For this PR it's limited to translating CMD+K to CTRL+K if you're on a non macOS/iOS platform, but once developers can integrate their plugins it'll be used to translate the shortcuts they use too.

File: war/src/main/js/keyboard-shortcuts.js
Comment (daniel-beck) [inline war/src/main/js/keyboard-shortcuts.js:23]: Are iPads iPhones?

Comment (janfaracik) [reply war/src/main/js/keyboard-shortcuts.js:23]: Just tested this on my devices (running Monterey and iOS/iPadOS 16) -

Mac: `MacIntel`
iPhone: `iPhone` (`MacIntel` when in desktop mode)
iPad: `MacIntel` (`iPad` when in mobile mode)

So not confusing at all 😭😂  

I've added an additional condition just in case a user is on mobile mode on their iPad.

Comment (daniel-beck) [reply war/src/main/js/keyboard-shortcuts.js:23]: > So not confusing at all



<details>
<summary>Wanna guess what's my `navigator.platform` on M1? </summary>

`MacIntel` lol

</details>

**Review by timja — APPROVED** [review]

File: war/src/main/js/keyboard-shortcuts.js
Comment (NotMyFault) [inline war/src/main/js/keyboard-shortcuts.js:25]: Looks like `replaceAll` doesn't work in test cases, https://ci.jenkins.io/job/Core/job/jenkins/job/PR-6893/7/consoleFull

Comment (timja) [reply war/src/main/js/keyboard-shortcuts.js:25]: I’ve hit that before you need to use replace I think

Comment (NotMyFault) [reply war/src/main/js/keyboard-shortcuts.js:25]: > I’ve hit that before you need to use replace I think

According to ES, it's part of 21, if we pull https://github.com/jenkinsci/jenkins/pull/6887 and merge it into this PR we should be good to go.

Comment (NotMyFault) [reply war/src/main/js/keyboard-shortcuts.js:25]: That didn't work out, does HtmlUnit not support ES6+ or anything modern released past 2015?

I guess we really need to use #replace again 😢

Comment (timja) [reply war/src/main/js/keyboard-shortcuts.js:25]: ha i doubted this would work, no htmlunit is far behind in support and it's a nightmare, which is why moving ATH to here and ditching htmlunit would be great to do at some point, but yes for now it needs to be replace

Comment (janfaracik) [reply war/src/main/js/keyboard-shortcuts.js]: Fingers crossed it's sorted now 🤞

Comment (NotMyFault) [reply war/src/main/js/keyboard-shortcuts.js]: > Fingers crossed it's sorted now 🤞

Build is green 👏🏻

**Comment by timja** [comment]
/label ready-for-merge

----

This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

**Comment by daniel-beck** [comment]
Nonblocking feedback:

> It's quite a common shortcut used across various services, such as GitHub and Spotify.

Those I'm aware of use it form some sort of context menu actions popup though. Seems weird to use it for something else here.

> I'm planning on using `/` for in page search fields, such as the filter build history input or the plugin manager search input.

WDYT about `/` for this search, `Alt + /` for context-specific search (or vice versa)?

**Comment by janfaracik** [comment]
> Those I'm aware of use it form some sort of context menu actions popup though. Seems weird to use it for something else here.
> 
> 
> 
> Although I guess not really, since it's a "launcher" more than a real search -- just the presentation differs from other sites. I guess this could go either way.

The eventual idea would be to adopt a similar launcher concept (not sure if you saw the UX meeting demo linked at the top). That's why I went for the CMD+K shortcut now just so it wouldn't change it the future/cause confusion.

image

**Comment by NotMyFault** [comment]
Causes [JENKINS-69250](https://issues.jenkins.io/browse/JENKINS-69250)

**Comment by daniel-beck** [comment]
@KalleOlaviNiemitalo Is this different from other sites? E.g. GitHub opens a popup on `Cmd-K`, so I would expect it to have the same issue.

**Pull request merged**
